parser/src/grammar.lalrpop

```diff
@@ -0,0 +1,144 @@
+}
+
+TraceColsGrp: TraceColsGrp = {
+    "main" ":" "[" <e: CommaElems<Expr1>> "]" => TraceColsGrp {
+        cols_grp_type: TraceColsGrpType::MainTraceCols,
+        cols: e,
+    } 
+}
+
+// TRANSITION CONSTRAINTS
+// ------------------------------------------------------------------------------------------------
+
+TransitionConstraints: TransitionConstraints = {
+    "transition_constraints" ":" <transition_constraints: Constraint+> =>
+        TransitionConstraints { transition_constraints }
+}
+
+// BOUNDARY CONSTRAINTS
+// ------------------------------------------------------------------------------------------------
+
+BoundaryConstraints: BoundaryConstraints = {
+    "boundary_constraints" ":" <boundary_constraints: Constraint+> =>
+        BoundaryConstraints { boundary_constraints }
+}
+
+Boundary: Boundary = {
+    "first" => Boundary::First,
+    "last" => Boundary::Last
+}
+
+// EXPRESSIONS WITH PRECEDENCE (LOWEST TO HIGHEST)
+// ------------------------------------------------------------------------------------------------
+
+Expr: Expr = {
+    Expr3
+}
+
+Expr3: Expr = {
+    <lexpr: Expr2> "=" <rexpr: Expr2> => Expr::Equal(Box::new(lexpr), Box::new(rexpr)),
+    Expr2
+}
+
+Expr2: Expr = {
+    <lexpr: Expr1> "+" <rexpr: Expr1> => Expr::Add(Box::new(lexpr), Box::new(rexpr)),
+    Expr1
+}
+
+Expr1: Expr = {
+    <s: Identifier> "." <boundary: Boundary> => Expr::Boundary(s, boundary),
+    <n: Num> => Expr::Int(Tok::Number(n.to_string())),
+    <s: Identifier> => Expr::Variable(s),
+}
+
+// ATOMS
+// ------------------------------------------------------------------------------------------------
+
+Constraint: Constraint = {
+    "enf" <expr: Expr> => Constraint { expr }
+}
+
+CommaElems<T>: Vec<T> = {
+    <e:T> <v:("," <T>)*> => {
+        let mut v = v;
+        v.insert(0, e);
+        v
+    }
+}
+
+Identifier: Identifier = {
+    <n:identifier> => Identifier{ name: n.to_string() }
+}
+
+Num: i64 = {
+    r"[0-9]+" =>? i64::from_str(&<>)
+        .map_err(|_| ParseError::User {
+            error: Error::ParseError(InvalidInt(format!("Int too big : {}", <>).to_string()))
+        })
+}
+
+// LEXER
+// ------------------------------------------------------------------------------------------------
+
+extern {
+    type Error = Error;
+    type Location = usize;
+
+    enum Tok {
+        identifier => Tok::Name(<String>),
+        r"[0-9]+" => Tok::Number(<String>),
+        "def" => Tok::Def,
+        "boundary_constraints" => Tok::BoundaryConstraints,
+        "trace_columns" => Tok::TraceColumnns,
+        "transition_constraints" => Tok::TransitionConstraints,
+        "main" => Tok::Main,
+        "enf" => Tok::Enf,
+        "first" => Tok::First,
+        "last" => Tok::Last,
+        ":" => Tok::Colon,
+        "," => Tok::Comma,
+        "=" => Tok::Equal,
+        "[" => Tok::Lsqb,
+        "]" => Tok::Rsqb,
+        "." => Tok::Dot,
+        "+" => Tok::Plus,
+    }
```

Comment:
nit: I would order as suggested elsewhere, or at least aim to have our order here be consistent with the order in the lexer and the order above to the extent that it's possible